wp-includes/functions.wp-service-workers.php

```diff
@@ -0,0 +1,78 @@
+<?php
+/**
+ * Service Worker functions.
+ *
+ * @since ?
+ *
+ * @package PWA
+ */
+
+/**
+ * Initialize $wp_service_workers if it has not been set.
+ *
+ * @global WP_Service_Workers $wp_service_workers
+ * @return WP_Service_Workers WP_Service_Workers instance.
+ */
+function wp_service_workers() {
+	global $wp_service_workers;
+	if ( ! ( $wp_service_workers instanceof WP_Service_Workers ) ) {
+		$wp_service_workers = new WP_Service_Workers();
+	}
+	return $wp_service_workers;
+}
+
+/**
+ * Register a new service worker.
+ *
+ * @since ?
+ *
+ * @param string $handle Name of the service worker. Should be unique.
+ * @param string $path   Relative path of the service worker.
+ * @param array  $deps   Optional. An array of registered script handles this depends on. Default empty array.
+ * @param string $scope  Optional Scope of the service worker.
+ * @return bool Whether the script has been registered. True on success, false on failure.
+ */
+function wp_register_service_worker( $handle, $path, $deps = array(), $scope = null ) {
+	$wp_service_workers = wp_service_workers();
+
+	// If the path is not relative. @todo should we try formatting it instead?
+	if ( ! preg_match( '/^\//', $path ) ) {
```

Comment:
The logic here can be mirrored from [`AMP_Style_Sanitizer::get_validated_url_file_path()`](https://github.com/Automattic/amp-wp/blob/70711f0cd647f233b02ffa1bbbd58496eeeea510/includes/sanitizers/class-amp-style-sanitizer.php#L441-L511), which is itself inspired by `\WP_Scripts::in_default_dir()` in core.